### ISSUE TITLE
__static yield absolute path

### ISSUE THREAD

**Issue by gxcodes** [SimulatedGREG/electron-vue#267]
#### On Minimize Milestone
First of all, congrats on this great milestone! I specifically migrate my electron app here to enjoy the __static var.

##### Describe the issue / bug.
I need to consume a basemap file for my map application. Unfortunately, I can not pass the path from the variable, as it returns an absolute path, the Drive is the root, not the application itself. Is this the intended behaviour? __static seems to yield different result, as compared to __dirname for example.

##### How can I reproduce this problem?
`const ps = path.join(__static, '/countries.mbtiles')`
`const pd = path.join(__dirname, '/countries.mbtiles')`
`console.log(ps) // returns file location in absolute path from drive, doesnt work`
`console.log(pd) // returns relative path, this works`

but the problem with __dirname is, on development I should put the file alongside my *.vue component, and it doesn't work in production as its packed in asar.

If I use the __static var, the path resulted was `http://localhost:9080/Users/myusername/myapp/static/countries.mbtiles` and its definitely not found. Shouldn't it be `http://localhost:9080/static/countries.mbtiles` during development?

##### Tell me about your development environment.
  * Node version: v8.1.0
  * NPM version: v5.0.3
  * vue-cli version: v2.8.2
  * Operating System: macOS Sierra

Really appreciate your time and effort. Thanks!

**Comment by SimulatedGREG** [comment]
@gxcodes 

The `__static` variable is intended to produce an absolute path to your project's `static/` directory. It is no way intended to be a replacement to `__dirname`, but is meant as an alternative to it. `__static` was solely created because `__dirname` does not produce an expected path within `webpack`, hence why you were getting a *relative* path. In ordinary `node`, `__dirname` would also produce an absolute path. `__static` is also only meant to be used with modules that need a path to a file/folder, such as `fs`.

> but the problem with __dirname is, on development I should put the file alongside my *.vue component, and it doesn't work in production as its packed in asar.

If you are **not** using `require`/`import`, `<img src="">`, or `@import` on your asset, then `webpack` **will never** bundle your asset in production. Any other assets that `webpack` doesn't have a loader for should be placed in `static/` and referenced with `__static`/`path`.

In short, `__static` is used to get a path to `static/`. This is a very helpful location to store `electron` specific assets such as icon sets that the `electron` APIs need a path to, such as [`nativeImage`](https://electron.atom.io/docs/api/native-image/).

Related: https://stackoverflow.com/questions/44337944/how-to-manage-configuration-for-webpack-electron-app/44379300#44379300

**Comment by gxcodes** [comment]
Thanks for the reply!

I wonder if there's any alternative to be able to pass this file to the function needed it. I tried another method like:
`const p = `file://${path.join(__static, '/countries.mbtiles')}``

But now it returns CORS related error because its not http anymore. Internally, the function (which I dont have access to modify) uses Fetch API.

Sorry if this is not related to electron-vue anymore, but maybe you can point me some direction before closing?

**Comment by SimulatedGREG** [comment]
What module are you trying to use? I'm a little confused as to how CORS is getting involved if you are dealing with the file system.

**Comment by gxcodes** [comment]
I'm making a electron app that display some maps and overlays on top of it. For the map library, I'm using [Leaflet](http://leafletjs.com/). I used a plugin to read an mbtiles, this works fine if I put the file somewhere 'online' but defeat the purpose on making the "desktop" version that's expected to be offline.

The plugin that needs to consume the mbtiles is [Leaflet.Tilelayer.MBTiles](https://gitlab.com/IvanSanchez/Leaflet.TileLayer.MBTiles)

**Comment by SimulatedGREG** [comment]
Looking at https://gitlab.com/IvanSanchez/Leaflet.TileLayer.MBTiles/blob/master/Leaflet.TileLayer.MBTiles.js#L82, it seems that a `<img>` tag is used internally, therefore using `__static` probably won't work. But since this is loading over browser protocol, you could try just using `static/countries.mbtitles` as the input.

**Comment by gxcodes** [comment]
Wow thanks! It totally worked during development. I guess I over-think everything.

But on production (with electron-builder), it doesn't show up, understandably because I guess now the renderer has no access to filesystem? Or I skipped something again?

**Comment by SimulatedGREG** [comment]
@gxcodes 

With the base electron-vue setup, using something like `static/myasset.png` on a `<img>` tag works both in development and production. Unless `leaflet` is doing something else internally to change how the asset is handled, I'm not sure why this wouldn't work in production. I noticed that this library also uses `sql.js` internally. Is there a chance your dev database is being used in production, therefore providing *unexpected* or *different* results.

**Comment by gxcodes** [comment]
So basically, mbtiles is an sqlite database that store tile images (usually png/jpg) that made the basemap. The plugin made sql.js as dependency to handle the corresponding tile request and forward them to Leaflet to be displayed.

I still scratched my head over this, it works really well in development, but displays nothing in production, not even error messages.

**Comment by SimulatedGREG** [comment]
@gxcodes 

![image](https://user-images.githubusercontent.com/8327811/27185942-7a978874-51ac-11e7-9578-a8524d6c543c.png)

The `fetch` API *intentionally* doesn't support the `file://` protocol. `electron` uses the `file://` protocol by default to load things from the file system. This only works in development because the use of `webpack-dev-server`, which of course uses the `http://` protocol that `fetch` was built for.

Luckily, `Leaflet.TileLayer.MBTiles` also supports providing an ArrayBuffer (https://gitlab.com/IvanSanchez/Leaflet.TileLayer.MBTiles/blob/master/Leaflet.TileLayer.MBTiles.js#L28). So instead of trying to load things over `fetch`, we can load `countries.mbtiles` into memory.

```js
import fs from 'fs'
import path from 'path'

const nodeBuffer = fs.readFileSync(path.join(__static, '/countries.mbtiles')).buffer
const offlineMap = L.tileLayer.mbTiles(nodeBuffer)
```

**BUT**, there's actually a bug in `Leaflet.TileLayer.MBTiles`. It is a quick fix, and I've also made an issue. (https://gitlab.com/IvanSanchez/Leaflet.TileLayer.MBTiles/issues/4)

After applying fix, using the above example, and switching the map to `Indonesia OFFLINE`,  the map loads in production with my Wifi off. 😄 

PS: You may not have seen any errors in production because `babili-webpack-plugin` is set to remove all console statements in production. (https://github.com/SimulatedGREG/electron-vue/blob/master/template/.electron-vue/webpack.renderer.config.js#L156)

**Comment by gxcodes** [comment]
You, sir, will have my eternal gratitude! It didn't cross my mind to check the plugin source-code. All is well now. Thanks again for this great boilerplate, I woudn't come this far without it.

PS: If you're ever in Indonesia, Jakarta area, ping me for a beer! :)
